Source/WebCore/accessibility/AccessibilityObject.cpp

```diff
@@ -1612,7 +1613,18 @@ std::optional<SimpleRange> AccessibilityObject::rangeForCharacterRange(const Cha
 VisiblePositionRange AccessibilityObject::lineRangeForPosition(const VisiblePosition& visiblePosition) const
 {
     VisiblePosition startPosition = startOfLine(visiblePosition);
-    VisiblePosition endPosition = nextLineEndPosition(startPosition);
+    VisiblePosition endPosition = endOfLine(visiblePosition);
+
+    if (endPosition.isNull() || endPosition < startPosition) {
+        // When endOfLine fails to return a plausible result, try nextLineEndPosition, which is more robust, but ensure it doesn't return a result from a subsequent line.
+        VisiblePosition nextLineEnd = nextLineEndPosition(startPosition);
+        while (!nextLineEnd.isNull() && nextLineEnd > startPosition && lineForPosition(nextLineEnd) != lineForPosition(startPosition))
```

Comment:
`lineForPosition` iterates every line until the nearest editable root, which could be really expensive, right? And we do that twice here. Would `inSameLine(nextLineEnd, startPosition)` work? Though one caveat is that `inSameLine` confusingly returns false if the positions are rendered on the same line, but one is editable (wrapped in `<span contenteditable="true">`) while the other isn't (e.g. wrapped by `<span contenteditable="false">`), and Google docs does have markup that can exhibit this (see the PR description of https://github.com/WebKit/WebKit/pull/30862).

Reply:
Great call, let me see if isSameLine works or not.

Reply:
inSameLine works great, done.